publisher/publisher.go

```diff
@@ -0,0 +1,274 @@
+	ctLogs            []LogDescription
+
+	SA core.StorageAuthority
+}
+
+// NewPublisherImpl creates a Publisher that will submit certificates
+// to any CT logs configured in CTConfig
+func NewPublisherImpl(ctConfig CTConfig) (pub PublisherImpl, err error) {
+	logger := blog.GetAuditLogger()
+	logger.Notice("Publisher Authority Starting")
+
+	if ctConfig.BundleFilename == "" {
+		err = fmt.Errorf("No CT submission bundle provided")
+		return
+	}
+	bundle, err := core.LoadCertBundle(ctConfig.BundleFilename)
+	if err != nil {
+		return
+	}
+	for _, cert := range bundle {
+		pub.issuerBundle = append(pub.issuerBundle, base64.StdEncoding.EncodeToString(cert.Raw))
+	}
+	ctBackoff, err := time.ParseDuration(ctConfig.SubmissionBackoffString)
+	if err != nil {
+		return
+	}
+
+	pub.log = logger
+	pub.client = &http.Client{}
+	pub.submissionBackoff = ctBackoff
+	pub.submissionRetries = ctConfig.SubmissionRetries
+	pub.ctLogs = ctConfig.Logs
+
+	return
+}
+
+func (pub *PublisherImpl) submitToCTLog(serial string, jsonSubmission []byte, log LogDescription) error {
+	var sct core.SignedCertificateTimestamp
+	backoff := pub.submissionBackoff
+	done := false
+	var retries int
+	for retries = 0; retries <= pub.submissionRetries; retries++ {
+		if retries > 0 {
+		}
+		resp, err := postJSON(pub.client, fmt.Sprintf("%s%s", log.URI, "/ct/v1/add-chain"), jsonSubmission, &sct)
+		if err != nil {
+			// Retry the request, log the error
+			// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+			pub.log.AuditErr(fmt.Errorf("Error POSTing JSON to CT log submission endpoint [%s]: %s", log.URI, err))
+			time.Sleep(backoff)
+			continue
+		} else {
+			if resp.StatusCode == http.StatusRequestTimeout || resp.StatusCode == http.StatusServiceUnavailable {
+				// Retry the request after either 10 seconds or the period specified
+				// by the Retry-After header
+				backoff = pub.submissionBackoff
+				if seconds, err := strconv.Atoi(resp.Header.Get("Retry-After")); err == nil {
+					backoff = time.Second * time.Duration(seconds)
+				} else {
+					fmt.Println(err)
+				}
+				time.Sleep(backoff)
+				continue
+			} else if resp.StatusCode != http.StatusOK {
+				// Not something we expect to happen, set error, break loop and log
+				// the error
+				// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+				pub.log.AuditErr(fmt.Errorf("Unexpected status code returned from CT log submission endpoint [%s]: Unexpected status code [%d]", log.URI, resp.StatusCode))
+				break
+			}
+			done = true
+			break
+		}
+	}
+
+	if !done {
+		pub.log.Warning(fmt.Sprintf(
+			"Unable to submit certificate to CT log [Serial: %s, Log URI: %s, Retries: %d]",
+			serial,
+			log.URI,
+			retries,
+		))
+		return fmt.Errorf("Unable to submit certificate")
+	}
+
+	if err := sct.CheckSignature(); err != nil {
+		// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+		pub.log.AuditErr(err)
+		return err
+	}
+
+	pub.log.Info(fmt.Sprintf(
```

Comment:
I would eliminate this log entry, or lower it to Debug level. It's redundant with the Info log below indicating that storage was successful.